pageserver/src/tenant/remote_timeline_client/index.rs

```diff
@@ -236,15 +260,53 @@ impl Lineage {
             .is_some_and(|(_, ancestor_lsn, _)| ancestor_lsn == lsn)
     }
 
-    pub(crate) fn is_detached_from_original_ancestor(&self) -> bool {
+    /// Returns true if the timeline originally had an ancestor, and no longer has one.
+    pub(crate) fn is_detached_from_ancestor(&self) -> bool {
         self.original_ancestor.is_some()
     }
 
+    /// Returns original ancestor timeline id and lsn that this timeline has been detached from.
+    pub(crate) fn detached_previous_ancestor(&self) -> Option<(TimelineId, Lsn)> {
+        self.original_ancestor.map(|(id, lsn, _)| (id, lsn))
+    }
+
     pub(crate) fn is_reparented(&self) -> bool {
         !self.reparenting_history.is_empty()
     }
 }
 
+/// Right now, the only reason to block gc persistently is detach_ancestor. To use gc blocking more
+/// broadly, a reason set field needs to be added, and the shared state load time building be
+/// complicated to avoid detach_ancestor clearing out a manually configured gc blocking.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub(crate) struct GcBlocking {
+    pub(crate) started_at: NaiveDateTime,
+}
+
+impl GcBlocking {
+    pub(super) fn started_now_for_detach_ancestor() -> Self {
+        GcBlocking {
+            started_at: chrono::Utc::now().naive_utc(),
+        }
+    }
+
+    /// Returns true if detach_ancestor is one of the reasons why the gc is blocked.
+    pub(crate) fn blocked_by_detach_ancestor(&self) -> bool {
+        true
+    }
+
+    /// Returns a version of self with the reason of detach_ancestor.
+    pub(super) fn with_detach_ancestor(&self) -> Self {
+        self.clone()
+    }
```

Comment:
> is already blocking because of something else?

I think it would just be:

```rust
let Self { started_at, reasons } = self;

assert!(reasons.contains(GcBlockingReason::DetachAncestor));

let reasons = {
  let mut reasons = reasons.clone();
  reasons.insert(GcBlockingReason::DetachAncestor);
  reasons
};

Self { started_at: *started_at, reasons }
```

I don't see why it should fail, but yeah, sure reasons could be added later..

Reply:
This has now been implemented #8600.